url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/pl/cc/Ping.java

```diff
@@ -0,0 +1,202 @@
+	 * @param interval czas w sekundach pomiďż˝dzy kolejnymi pingami
+	 * @param timeOut tyle sek moze upďż˝ynďż˝ďż˝ od wpďż˝yniďż˝cia ostatniego ponga
+	 */
+	public Ping(ProxyEventListenerList proxyEventListenerList, NetPrintWriter out, int interval, int timeOut){
+		this(proxyEventListenerList, out);
+		this.interval = interval;
+		this.timeOut = timeOut;
+	}
+	
+	private void checkPong(){
+		if ((Utils.getNow().getTime()-lastPong.getTime())>(timeOut*1000)){
+			log.error("Brak ďż˝ďż˝cznoďż˝ci z ProxyAgent");
+			proxyEventListenerList.onProblemWithPing();
+			stop = true;
+		}
+	}
+
+	public void run(){
+		setName("ping");
+		log.info("Ping thread started");
+		checkPingThread = new CheckPingThread();
+		checkPingThread.start();
+		while (!stop){
+			try {
+				lastPing = Utils.getNow();
+				log.debug("Send PING: "+lastPing.toString());
+				out.println("ping");
+				sleep(interval*1000);
+				if (stop) break;
+			} catch (NullPointerException e1) {
+				log.warn("Problem z ping-pong (NullPointerException)");
+			} catch (InterruptedException e2) {
+				log.warn("Problem z ping-pong (InterruptedException)");
+			}
+		}
+		log.debug("Koniec działania wątku ping");
+	}
+	
+	public void stopThread() {
+		stop = true;
+	}
+
+	public void setLastPong(Date date){
+		lastPong = date;
+		log.debug("Recieve PONG: "+date.toString());
+	}
+	
+	public Date getLastPong() {
+		return lastPong;
+	}
+
+	@Override
+	public void onAuthorized(OkWelcome selfInfo) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onAuthResponse(boolean authorized) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onDisconnect() {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onEvent(Command command) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onInfoVersion(InfoVersion infoVersion) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onPong(Date now) {
+		setLastPong(now);
+	}
+
+	@Override
+	public void onProblemWithPing() {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onTryConnect(boolean successfully, String host) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onWelcome() {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onInvalidVersion(String applicationVersion, InfoVersion infoVersion) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onStartReconnectProcedure(int sleepBeforeReconnect) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void onConnectionReplaced(ErrSameAgentConnection connectionReplaced) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	
+	class CheckPingThread extends Thread {
+			public void run(){
+				setName("CheckPingThread");
+				while (true){
+					if (stop) return;
+					checkPong();
+					try {
+						Thread.sleep(interval*1000);
+					} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)